src/engine/guidance/collapse_scenario_detection.cpp

```diff
@@ -0,0 +1,404 @@
+#include "engine/guidance/collapse_scenario_detection.hpp"
+#include "extractor/guidance/constants.hpp"
+#include "util/bearing.hpp"
+
+#include <numeric>
+
+#include <boost/assert.hpp>
+
+namespace osrm
+{
+namespace engine
+{
+namespace guidance
+{
+
+namespace
+{
+
+// check bearings for u-turns.
+// since bearings are wrapped around at 0 (we only support 0,360), we need to do some minor math to
+// check if bearings `a` and `b` go in opposite directions. In general we accept some minor
+// deviations for u-turns.
+bool bearingsAreReversed(const double bearing_in, const double bearing_out)
+{
+    // Nearly perfectly reversed angles have a difference close to 180 degrees (straight)
+    const double left_turn_angle = [&]() {
+        if (0 <= bearing_out && bearing_out <= bearing_in)
+            return bearing_in - bearing_out;
+        return bearing_in + 360 - bearing_out;
+    }();
+    return util::angularDeviation(left_turn_angle, 180) <= 35;
+}
+
+// to collapse steps, we focus on short segments that don't interact with other roads. To collapse
+// two instructions into one, we need to look at to instrutions immediately after each other.
+bool noIntermediaryIntersections(const RouteStep &step)
+{
+    return std::all_of(step.intersections.begin() + 1,
+                       step.intersections.end(),
+                       [](const auto &intersection) { return intersection.entry.size() == 2; });
+}
+
+// Link roads, as far as we are concerned, are short unnamed segments between to named segments.
+bool isLinkroad(const RouteStep &pre_link_step,
+                const RouteStep &link_step,
+                const RouteStep &post_link_step)
+{
+    const constexpr double MAX_LINK_ROAD_LENGTH = 2 * MAX_COLLAPSE_DISTANCE;
+    const auto is_short = link_step.distance <= MAX_LINK_ROAD_LENGTH;
+    const auto unnamed = link_step.name_id == EMPTY_NAMEID;
+    const auto between_named =
+        (pre_link_step.name_id != EMPTY_NAMEID) && (post_link_step.name_id != EMPTY_NAMEID);
+
+    return is_short && unnamed && between_named && noIntermediaryIntersections(link_step);
+}
+
+// Just like a link step, but shorter and no name restrictions.
+bool isCollapsableSegment(const RouteStep &step)
```

Comment:
Carrying over from https://github.com/Project-OSRM/osrm-backend/pull/3705#pullrequestreview-21548743, the name of this function should be made a bit clearer.